WcaOnRails/app/models/result.rb

```diff
@@ -1,15 +1,85 @@
 # frozen_string_literal: true
 class Result < ActiveRecord::Base
+  include ResultMethods
+
   self.table_name = "Results"
 
   belongs_to :competition, foreign_key: :competitionId
+  belongs_to :country, foreign_key: :countryId
   belongs_to :person, -> { current }, primary_key: :wca_id, foreign_key: :personId
   belongs_to :round, foreign_key: :roundId
   belongs_to :event, foreign_key: :eventId
+  belongs_to :format, foreign_key: :formatId
 
   scope :podium, -> { joins(:round).merge(Round.final_rounds).where(pos: [1..3]).where("best > 0") }
   scope :winners, -> { joins(:round, :event).merge(Round.final_rounds).where("pos = 1 and best > 0").order("Events.rank") }
 
+  validate :validate_each_solve
+  def validate_each_solve
+    solve_times.each_with_index do |solve_time, i|
+      unless solve_time.valid?
+        errors.add(:"value#{i + 1}", solve_time.errors.messages[:base].join(" "))
+      end
+    end
+  end
+
+  validate :validate_best
+  def validate_best
+    correct_best_solve_time = sorted_solves.first
+    if correct_best_solve_time && correct_best_solve_time.wca_value != best
+      errors.add(:best, "should be #{correct_best_solve_time.wca_value}")
+    end
+  end
+
+  validate :validate_number_of_solves
+  def validate_number_of_solves
+    return errors.add(:competitionId, "invalid") unless competition
+    return errors.add(:countryId, "invalid") unless country
+    return errors.add(:roundId, "invalid") unless round
+    return errors.add(:eventId, "invalid") unless event
+    return errors.add(:formatId, "invalid") unless format
+    return errors.add(:base, "Cannot skip all solves.") if solve_times.all?(&:skipped?)
+
+    previous_solve_time = solve_times.first
+    solve_times[1..-1].each do |solve_time|
+      if !solve_time.skipped? && previous_solve_time.skipped?
+        return errors.add(:base, "Skipped solves must all come at the end.")
+      end
+      previous_solve_time = solve_time
+    end
+
+    unskipped_count = solve_times.length - solve_times.count(&:skipped?)
+    if round.combined?
+      if unskipped_count > format.expected_solve_count
+        return errors.add(:base, "Expected at most #{format.expected_solve_count} #{'solve'.pluralize(format.expected_solve_count)}, but found #{unskipped_count}.")
+      end
+    else
+      if unskipped_count != format.expected_solve_count
+        return errors.add(:base, "Expected #{format.expected_solve_count} #{'solve'.pluralize(format.expected_solve_count)}, but found #{unskipped_count}.")
```

Comment:
I realized why these `pluralize` calls look so weird to me. Apparently, there are two `pluralize` functions in Rails!  The other one is better: With http://apidock.com/rails/ActionView/Helpers/TextHelper/pluralize, this

`#{format.expected_solve_count} #{'solve'.pluralize(format.expected_solve_count)}`

becomes this

`#{pluralize(format.expected_solve_count, 'solve')}`

I put this in my models when I need helper functions:

    def hlp
      ActionController::Base.helpers
    end